### DISCUSSION TITLE
Using Generics to provide Type Safety in Zod Object Shape

### DISCUSSION THREAD

**Discussion by TheRealFlyingCoder** [colinhacks/zod#1900] [Q&A] [answered]
CONTEXT

1. You have an external interface full of nullable values (or something similar) <-- You know the object structure

```ts
interface Nullables {
  one?: string | null;
  two?: string | null;
}
```

2. All the fields are actually required, so you use zod for validation: <-- Structure must stay the same

```ts
const schema = z.object({
  one: z.string(),
  two: z.string()
})
```

QUESTION

Given this external interface could be updated at any time, how do you use it as a generic, so that the zod object shape can be typesafe?

i.e:

```ts
const schema = z.object<Nullables>({
  one: z.string(),
  two: z.number() //TS Error
  three: z.string() //TS rror
})
```

If we already have the object structure available before typing out the zod object, it would be very handy to use it so that we can be made aware of any mistakes, or changes automatically.

**Comment by JacobWeisenburger** [comment, marked as answer]
This is the best way that I know of.
https://github.com/colinhacks/zod/discussions/1886#discussioncomment-4686267

**Reply by TheRealFlyingCoder** [reply]
Yeah but the problem here, is that you don't actually have any validation of the shape.

I could have a completely different zod object and it would satisfy the type.

All this does is ensure that for any given keys on the object, my zod return type matches:

```
one: z.string() //Works

one: z.number() //TS Error
```

I can pass `satisfies z.ZodType<Required<Nullable>>` though which makes sure I have at least all the keys on my object.

But regardless of that there is still no intellisense or support gained through generics

**Reply by TheRealFlyingCoder** [reply]
But thanks for pointing this out it's at least one step closer haha
